src/engine/SCons/CacheDir.py

```diff
@@ -145,56 +148,91 @@ def __init__(self, path):
+                # safe for multiple processes creating the same directory,
+                # apart from the last entry.
+                os.makedirs(path)
+            except OSError:
+                if not os.path.isdir(path):
+                    msg = "Failed to create cache directory " + path
+                    raise SCons.Errors.EnvironmentError(msg)
+
         # See if there's a config file in the cache directory. If there is,
         # use it. If there isn't, and the directory exists and isn't empty,
         # produce a warning. If the directory doesn't exist or is empty,
         # write a config file.
         config_file = os.path.join(path, 'config')
-        if not os.path.exists(config_file):
-            # A note: There is a race hazard here, if two processes start and
-            # attempt to create the cache directory at the same time. However,
-            # python doesn't really give you the option to do exclusive file
-            # creation (it doesn't even give you the option to error on opening
-            # an existing file for writing...). The ordering of events here
-            # as an attempt to alleviate this, on the basis that it's a pretty
-            # unlikely occurence (it'd require two builds with a brand new cache
-            # directory)
-            if os.path.isdir(path) and len(os.listdir(path)) != 0:
-                self.config['prefix_len'] = 1
-                # When building the project I was testing this on, the warning
-                # was output over 20 times. That seems excessive
-                global warned
-                if self.path not in warned:
-                    msg = "Please upgrade your cache by running " +\
-                          " scons-configure-cache.py " +  self.path
-                    SCons.Warnings.warn(SCons.Warnings.CacheVersionWarning, msg)
-                    warned[self.path] = True
-            else:
-                if not os.path.isdir(path):
-                    try:
-                        os.makedirs(path)
-                    except OSError:
-                        # If someone else is trying to create the directory at
-                        # the same time as me, bad things will happen
-                        msg = "Failed to create cache directory " + path
-                        raise SCons.Errors.EnvironmentError(msg)
-                        
-                self.config['prefix_len'] = 2
+        tmp_base = 'config.tmp'
+        tmp_file = os.path.join(path, tmp_base)
+        lock_base = 'config.lock'
+        lock_file = os.path.join(path, lock_base)
+
+        cache_create_tries = 0
+        while not os.path.exists(config_file):
+            # Lock the whole process of creating the directory and config file.
+            try:
+                os.remove(lock_file)
+            except OSError:
+                pass
+            try:
+                lock = os.open(lock_file, os.O_CREAT | os.O_EXCL)
```

Comment:
Relatively convinced this is not going to work. On Linux/UNIX  os.remove will remove the directory entry for a file, even if the file remains open.  Then the open + create + exclusive will complete happily, because open works off the directory entry. So now you have two processes thinking they hold the lock file.